crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs

```diff
@@ -88,48 +92,72 @@ pub(super) async fn configure_dsn(
-                            }
+        request_response_protocols: vec![
+            PieceByHashRequestHandler::create(move |req| {
+                let result = {
+                    debug!(piece_index_hash = ?req.piece_index_hash, "Piece request received. Trying cache...");
+                    let multihash = req.piece_index_hash.to_multihash();
+
+                    let piece_from_cache = piece_store.get(&multihash.into());
+
+                    if piece_from_cache.is_some() {
+                        piece_from_cache
+                    } else {
+                        debug!(piece_index_hash = ?req.piece_index_hash, "No piece in the cache. Trying archival storage...");
+
+                        let read_piece_fut = {
+                            let readers_and_pieces = match weak_readers_and_pieces.upgrade() {
+                                Some(readers_and_pieces) => readers_and_pieces,
+                                None => {
+                                    debug!("A readers and pieces are already dropped");
+                                    return None;
+                                }
+                            };
+                            let readers_and_pieces = readers_and_pieces.lock();
+                            let readers_and_pieces = match readers_and_pieces.as_ref() {
+                                Some(readers_and_pieces) => readers_and_pieces,
+                                None => {
+                                    debug!(
+                                        ?req.piece_index_hash,
+                                        "Readers and pieces are not initialized yet"
+                                    );
+                                    return None;
+                                }
+                            };
+
+                            readers_and_pieces.read_piece(&req.piece_index_hash)?
                         };
 
-                        readers_and_pieces.read_piece(&req.piece_index_hash)?
-                    };
+                        let handle = piece_protocol_handle.clone();
+                        tokio::task::block_in_place(move || handle.block_on(read_piece_fut))
+                    }
+                };
 
-                    let handle = handle.clone();
-                    tokio::task::block_in_place(move || handle.block_on(read_piece_fut))
-                }
-            };
+                Some(PieceByHashResponse { piece: result })
+            }),
+            RootBlockBySegmentIndexesRequestHandler::create(move |req| {
+                debug!(segment_indexes_count = ?req.segment_indexes.len(), "Root blocks request received.");
+
+                let handle = root_blocks_protocol_handle.clone();
+                let node_client = node_client.clone();
+                let internal_result = tokio::task::block_in_place(move || {
+                    handle.block_on(node_client.root_blocks(req.segment_indexes.clone()))
```

Comment:
We'll need to do something about these eventually, it is not great that we're blocking executor like this, especially when making RPC requests over the network.